shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -517,4 +475,36 @@ public boolean performEditorAction(int actionCode) {
     }
     return true;
   }
+
+  // -------- Start: ListenableEditingState watcher implementation -------
+  @Override
+  public void didChangeEditingState(
+      boolean textChanged, boolean selectionChanged, boolean composingRegionChanged) {
+    // Notifies the input method that the editing state has changed. The input method may decide
+    // to modify/reject these changes and ends up calling TextInputChannel#updateEditingState to
+    // send the update to the framework. This could result in an infinite loop if the framework
+    // also rejects/modifies the new update.
+
+    // Always send selection update. InputMethodManager#updateSelection skips sending the message
+    // if none of the parameters have changed since the last time we called it.
+    mImm.updateSelection(
+        mFlutterView,
+        mEditable.getSelecionStart(),
+        mEditable.getSelecionEnd(),
+        mEditable.getComposingStart(),
+        mEditable.getComposingEnd());
+
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
+      return;
+    }
+    if (mExtractRequest != null) {
+      mImm.updateExtractedText(
```

Comment:
Added comments for `getExtractedText`. In landscape mode some IMEs enter fullscreen mode (and switch to their own editors) to make the best use of the limited vertical space. `getExtractedText` is meant for that mode. Samsung keypad seems to use that instead of `getText{Before,After}Selection` to query the current editing state (according to the documentation it seems to be reasonable because by using `getExtractedText`'s monitor flag you can avoid the IPC round trip https://developer.android.com/reference/android/view/inputmethod/InputConnection#getExtractedText(android.view.inputmethod.ExtractedTextRequest,%20int)).

Reply:
Thanks for the explanation and comments, I didn't know that.